### PR TITLE
Master imp add formula vectorization laa

### PR DESCRIPTION
**Pull request by laa-odoo**
## Concerning performances:

I don't measure any significant change for the formulas dataset and for the array formulas dataset.

Regarding the addition of vectorized formulas, a large dataset was created too.
It has exactly the same data as in the formulas data set. See the difference above : 

| makeLargeDataset                    | Evaluate All Cells | Model Created In |
|---------------------|--------------------|------------------|
| **formulas**        | ~ 270 ms                | ~ 2000 ms             |
| **vectorizedFormulas** | ~ 350 ms                | ~ 380 ms              |

Task: : [3356482](https://www.odoo.com/web#id=3356482&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/formulas/compiler.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ const OPERATOR_MAP = {
   "<=": "LTE",
   "<": "LT",
   "^": "POWER",
-  "&": "CONCATENATE",
+  "&": "CONCAT",
```
Comment (rrahir) [inline src/formulas/compiler.ts:25]: What is the concrete difference of behaviour of the 2 functions? (concat vs concatenate)

Comment (laa-odoo) [reply src/formulas/compiler.ts:25]: `CONCATENATE` takes ranges and not single arguments as input. Vectorization is only possible with simple arguments.

Besides, that's probably why both exist, they both do the same thing, but one will work as an aggregator and the other will vectorize

File: src/functions/index.ts [hunk 2]
```diff
@@ -121,10 +152,37 @@ function addInputHandling(
...
+            return arg;
+        }
+      });
+
+    return generateMatrix(countVectorizableCol, countVectorizableRow, (col, row) => {
+      if (col > vectorizableColLimit - 1 || row > vectorizableRowLimit - 1) {
+        return notAvailableError;
+      }
+      const matrixElm = compute.apply(this, argsVector(col, row));
+      // to avoid mismatching behavior when user try to vectorize arguments on formula that spread
```
Comment (LucasLefevre) [inline src/functions/index.ts:179]: I'm not sure to understand what this means

File: src/functions/index.ts [hunk 3]
```diff
@@ -83,33 +83,64 @@ export class FunctionRegistry extends Registry<FunctionDescription> {
...
       if (isMatrix(arg) && !argDefinition.acceptMatrix) {
-        if (arg.length !== 1 || arg[0].length !== 1) {
-          throw new EvaluationError(
-            _t(
-              "Function [[FUNCTION_NAME]] expects the parameter '%s' to be a single value or a single cell reference, not a range.",
-              argDefinition.name
-            )
-          );
+        // if argDefinition does not accept a matrix but arg is still a matrix
+        // --> triggers the vectorization of the arguments
```
Comment (LucasLefevre) [inline src/functions/index.ts:123]:
```suggestion
        // --> triggers the arguments vectorization
```

File: src/formulas/compiler.ts (hunk 1 above)
Comment (LucasLefevre) [inline src/formulas/compiler.ts:25]: maybe a little test ? :roll_eyes:

File: src/functions/index.ts [hunk 4]
```diff
@@ -83,33 +83,64 @@ export class FunctionRegistry extends Registry<FunctionDescription> {
...
+          vectorArgsType.push("horizontal");
+          countVectorizableCol = Math.max(countVectorizableCol, arg.length);
+          vectorizableColLimit = Math.min(vectorizableColLimit, arg.length);
+        } else if (arg[0].length !== 1) {
+          vectorArgsType.push("vertical");
+          countVectorizableRow = Math.max(countVectorizableRow, arg[0].length);
+          vectorizableRowLimit = Math.min(vectorizableRowLimit, arg[0].length);
+        } else {
+          vectorArgsType.push("scalar");
+          args[i] = arg[0][0];
```
[the remark this answers is not in the document]
Comment (anhe-odoo) [reply src/functions/index.ts:140]: No, we need to update it here as, if we only have 1x1 elements, we won't go through the `argsVector` function as `countVectorizableCol == countVectorizableRow == 1`, but we need to have only scalar values in the args then.

File: src/functions/index.ts [hunk 5]
```diff
@@ -83,33 +83,64 @@ export class FunctionRegistry extends Registry<FunctionDescription> {
...
         }
-        args[i] = arg[0][0];
+      } else {
+        vectorArgsType.push("scalar");
```
[the remark this answers is not in the document]
Comment (anhe-odoo) [reply src/functions/index.ts:143]: What about something like this : 
```js
let vectorArgsType: VectorArgType[] | undefined = undefined;
for (let i = 0; i < args.length; i++) {
  const argDefinition = descr.args[descr.getArgToFocus(i + 1) - 1];
  const arg = args[i];

  if (isMatrix(arg) && !argDefinition.acceptMatrix) {
    // if argDefinition does not accept a matrix but arg is still a matrix
    // --> triggers the arguments vectorization
    const nColumns = arg.length;
    const nRows = arg[0].length;
    if (nColumns !== 1 || nRows !== 1) {
      vectorArgsType ??= new Array(args.length);
      if (nColumns !== 1 && nRows !== 1) {
        vectorArgsType[i] = "matrix";
        countVectorizableCol = Math.max(countVectorizableCol, nColumns);
        countVectorizableRow = Math.max(countVectorizableRow, nRows);
        vectorizableColLimit = Math.min(vectorizableColLimit, nColumns);
        vectorizableRowLimit = Math.min(vectorizableRowLimit, nRows);
      } else if (nColumns !== 1) {
        vectorArgsType[i] = "horizontal";
        countVectorizableCol = Math.max(countVectorizableCol, nColumns);
        vectorizableColLimit = Math.min(vectorizableColLimit, nColumns);
      } else if (nRows !== 1) {
        vectorArgsType[i] = "vertical";
        countVectorizableRow = Math.max(countVectorizableRow, nRows);
        vectorizableRowLimit = Math.min(vectorizableRowLimit, nRows);
      }
    } else {
      args[i] = arg[0][0];
    }
  }
```

Then, if we only have 1x1 elements (like in the base case, where we use reference to cell as parameters), we won't create the arrray too. The switch can stay as you proposed : 
```js
const argsVector: (i: number, j: number) => Arg[] = (i, j) =>
  args.map((arg, index) => {
    switch (vectorArgsType?.[index]) {
      case "matrix":
        return arg![i][j];
      case "horizontal":
        return arg![i][0];
      case "vertical":
        return arg![0][j];
      case undefined:
        return arg;
    }
  });
```

File: src/functions/index.ts (hunk 2 above)
Comment (anhe-odoo) [reply src/functions/index.ts:179]: // In the case where the user tries to vectorize arguments of an array formula, we will get an
// array for every combination of the vectorized arguments, which will lead to a 3D matrix and
// we won't be able to return the values.
// In this case, we keep the first element of each spreading part, just as Excel does, and
// create an array with these parts.
// For exemple, we have MUNIT(x) that return an unitary matrix of x*x. If we use it with a
// range, like MUNIT(A1:A2), we will get two unitary matrixes (one for the value in A1 and one
// for the value in A2). In this case, we will simply take the first value of each matrix and
// return the array [First value of MUNIT(A1), First value of MUNIT(A2)].

File: src/functions/index.ts [hunk 6]
```diff
@@ -83,33 +83,65 @@ export class FunctionRegistry extends Registry<FunctionDescription> {
...
+    this.mapping[name] = addErrorHandling(
+      addInputHandling(descr, addResultHandling(descr.compute, name)),
+      name
+    );
     super.add(name, descr);
     return this;
   }
 }
 
+type VectorArgType = "scalar" | "horizontal" | "vertical" | "matrix";
```
Comment (LucasLefevre) [inline src/functions/index.ts:95]: remove scalar

File: tests/functions/vectorization.test.ts [hunk 7]
```diff
@@ -121,4 +122,24 @@ describe("vectorization", () => {
...
+        functionDefinition.args.every((arg) =>
+          arg.type.every((t) => !t.startsWith("RANGE") && t !== "META")
+        )
+      );
+    }
+  });
+
+  test("unary operators should always accept vectors", () => {
+    // mean unary operators args should always be simple arg
+    for (let op in UNARY_OPERATOR_MAP) {
```
Comment (LucasLefevre) [inline tests/functions/vectorization.test.ts:140]: const

File: src/functions/index.ts [hunk 8]
```diff
@@ -121,10 +153,44 @@ function addInputHandling(
...
+      // In the case where the user tries to vectorize arguments of an array formula, we will get an
+      // array for every combination of the vectorized arguments, which will lead to a 3D matrix and
+      // we won't be able to return the values.
+      // In this case, we keep the first element of each spreading part, just as Excel does, and
+      // create an array with these parts.
+      // For exemple, we have MUNIT(x) that return an unitary matrix of x*x. If we use it with a
+      // range, like MUNIT(A1:A2), we will get two unitary matrixes (one for the value in A1 and one
+      // for the value in A2). In this case, we will simply take the first value of each matrix and
+      // return the array [First value of MUNIT(A1), First value of MUNIT(A2)].
+      return isMatrix(matrixElm) ? matrixElm[0][0] : matrixElm;
```
Comment (VincentSchippefilt) [inline src/functions/index.ts:189]: maybe add a test for this case. Coverage shows matrixElm[0][0] is never executed

**Review by VincentSchippefilt — APPROVED** [review]
I hereby officially approve merging the vectorisation.

**Review by LucasLefevre — APPROVED** [review]
yougotthis-martini

robodoo r+ rebase-ff

**Pull request closed without merging**